[SourceKit] Cancel in-flight builds on `editor.close`

hamishknight opened the pull request:
When closing a document, cancel any in-flight builds happening for it.

rdar://127126348

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please build toolchain

hamishknight commented on unittests/SourceKit/SwiftLang/CloseTest.cpp:
```diff
@@ -0,0 +1,178 @@
+  auto *Contents = getComplexSourceText();
+  const char *Args[] = {"-parse-as-library"};
+
+  for (auto RemoveCache : {true, false, false}) {
```
Heh, I meant to add a comment explaining but never did. It's to test the prior state for both the ASTProducer being cached and not cached

hamishknight replied:
Added a comment for this

bnbarham commented on tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp:
```diff
@@ -2444,8 +2455,14 @@ void SwiftLangSupport::editorClose(StringRef Name, bool RemoveCache) {
+    // Cancel any in-flight builds for the given AST.
+    Removed->cancelBuildsForCachedAST();
+
+    // Then remove the cached AST if we've been asked to do so.
+    if (RemoveCache)
+      Removed->removeCachedAST();
```
Is there any point to not removing from the cache if we never re-use it anyway? Or is there some case where it is re-used?

hamishknight committed:
[test] NFC: Extract out `NullEditorConsumer`

hamishknight committed:
[test] Disable syntax map for NullEditorConsumer

This doesn't seem to be required by any of the
tests using it.

hamishknight committed:
[SourceKit] Ensure `ASTCache` is guarded by mutex

There were a couple of accesses not guarded by
`CacheMtx`, introduce a couple of methods that
guard them, renaming `getASTProducer` while here.

Also make sure we don't ever insert a producer
after it has been purposefully removed by e.g a
close that removes the cached AST.

hamishknight committed:
[SourceKit] Cancel in-flight builds on `editor.close`

When closing a document, cancel any in-flight
builds happening for it.

rdar://127126348

hamishknight committed:
[test] Switch couple of SourceKit tests to `INITIALIZE_LLVM`

And move this into the constructor since it only
needs doing once.

hamishknight committed:
[SourceKit] Introduce key for cancelling on close

Defaults to `true`.

hamishknight commented:
@swift-ci please test

hamishknight commented:
After talking with @ahoppen, updated to use an opt-in flag for this behavior

hamishknight commented:
@swift-ci please build toolchain

hamishknight commented:
After further discussion, flipped the default back to `true`

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please build toolchain
